Refactor GraphDBSparqlService updateGraph and add RdfSparqlService

alexey-lukashov opened the pull request:
Refactor GraphDBSparqlService updateGraph and add RdfSparqlService implementation

- Made updateGraph in GraphDBSparqlService independent as required by SparqlService default methods
- Removed private uploadTtl to allow use of default implementation
- Added GraphDB-based implementation for RdfSparqlService
- Added single transaction replaceGraph to both GraphDB implementations

See issue #26

alexey-lukashov committed:
Refactor GraphDBSparqlService updateGraph and add RdfSparqlService implementation

- Made updateGraph in GraphDBSparqlService independent as required by SparqlService default methods
- Removed private uploadTtl to allow use of default implementation
- Added GraphDB-based implementation for RdfSparqlService

natan-cox-cognizone commented on semanticz-connector-graphdb/src/main/java/zone/cogni/semanticz/connectors/graphdb/GraphDBRdfStoreService.java:
```diff
@@ -0,0 +1,215 @@
+
+    @Override
+    public void close() {
+        RdfStoreService.super.close();
```
weird. super.close() is not possible?

alexey-lukashov replied:
nope, RdfStoreService is an interface with default method close, it does almost nothing just prints info log

alexey-lukashov replied:
super.close(); tries to resolve method close from superclass which is Object

natan-cox-cognizone commented on semanticz-connector-graphdb/src/main/java/zone/cogni/semanticz/connectors/graphdb/GraphDBSparqlService.java:
```diff
@@ -1,125 +1,227 @@
-/*
```
why are these deleted? should they not always be added? same for other files

alexey-lukashov replied:
yes, will fix this

alexey-lukashov committed:
fix headers

alexey-lukashov committed:
super.close

psiotwo commented on semanticz-connector-graphdb/src/main/java/zone/cogni/semanticz/connectors/graphdb/GraphDBSparqlService.java:
```diff
@@ -19,107 +19,225 @@
-    //  HttpClientBuilder httpClientBuilder = HttpClients.custom().useSystemProperties();
-    httpClient = HttpClientUtils.createHttpClientBuilder(config.getUser(), config.getPassword()).build();
-  }
+    private synchronized HttpClient getHttpClient() {
```
Why not to have this logic in HttpClientUtils?

psiotwo commented on semanticz-connector-graphdb/src/main/java/zone/cogni/semanticz/connectors/graphdb/GraphDBSparqlService.java:
```diff
@@ -19,107 +19,225 @@
+        if (httpClient != null) return httpClient;
+
+        HttpClient.Builder builder = HttpClient.newBuilder()
+                .connectTimeout(Duration.ofSeconds(5))
```
This should be configurable ideally.

psiotwo commented on semanticz-connector-graphdb/src/main/java/zone/cogni/semanticz/connectors/graphdb/GraphDBSparqlService.java:
```diff
@@ -19,107 +19,225 @@
+        HttpClient.Builder builder = HttpClient.newBuilder()
+                .connectTimeout(Duration.ofSeconds(5))
+                .followRedirects(HttpClient.Redirect.NORMAL)
+                .proxy(ProxySelector.getDefault());
```
Isn't this a default behaviour anyway?

psiotwo commented on semanticz-connector-graphdb/src/main/java/zone/cogni/semanticz/connectors/graphdb/GraphDBSparqlService.java:
```diff
@@ -19,107 +19,225 @@
+
+    @Override
+    public <R> R executeSelectQuery(String query, Function<ResultSet, R> resultHandler) {
+        try (QueryExecution queryExecution = QueryExecutionHTTP.service(config.getSparqlEndpoint())
```
Creation of the query execution is common for SELECT + ASK.

psiotwo commented on semanticz-connector-graphdb/src/main/java/zone/cogni/semanticz/connectors/graphdb/GraphDBSparqlService.java:
```diff
@@ -19,107 +19,225 @@
+                .POST(HttpRequest.BodyPublishers.ofString(turtle, StandardCharsets.UTF_8))
+                .build();
+
+        try {
```
There are significant common parts to replaceGraph and updateGraph - would be good to have a common method for them.

psiotwo commented on semanticz-connector-graphdb/src/main/java/zone/cogni/semanticz/connectors/graphdb/GraphDBRdfStoreService.java:
```diff
@@ -0,0 +1,234 @@
+        model.write(writer, "ttl");
+        String turtle = writer.toString();
+
+        String encodedCtx = URLEncoder.encode("<" + graphUri + ">", StandardCharsets.UTF_8);
```
addData and replaceGraph share quite some common code.

psiotwo reviewed (commented):
see comments

alexey-lukashov committed:
- Move HttpClient creation logic to HttpClientUtils with configurable options
  - Add connectTimeoutSeconds configuration to GraphDBConfig (default: 5s)
  - Extract common query execution creation for SELECT and ASK queries
  - Refactor to eliminate code duplication between replaceGraph and updateGraph in GraphDBSparqlService
  - Refactor to eliminate code duplication between addData and replaceGraph in GraphDBRdfStoreService
  - Extract HTTP method constants to avoid duplicate string literals
  - Make proxy selector usage configurable through HttpClientUtils
